scripts/hotWalletCli/client.ts

```diff
@@ -0,0 +1,283 @@
+      })
+      const totalRewardUnits = await contract.read.earned([address], { blockNumber: endBlock })
+
+      const rewardUnits = totalRewardUnits - totalRewardUnitsPrevEpoch
+
+      closingStateByStakingAddress[address] = { rewardUnits, totalRewardUnits, runeAddress }
+    }
+
+    return closingStateByStakingAddress
+  }
+
+  private async getDistributionsByStakingAddress(
+    closingStateByStakingAddress: ClosingStateByStakingAddress,
+    totalDistribution: BigNumber,
+  ) {
+    const totalEpochRewardUnits = Object.values(closingStateByStakingAddress).reduce(
+      (prev, { rewardUnits }) => prev + rewardUnits,
+      0n,
+    )
+
+    const distributionsByStakingAddress: Record<string, RewardDistribution> = {}
+    for await (const [address, { rewardUnits, totalRewardUnits, runeAddress }] of Object.entries(
+      closingStateByStakingAddress,
+    )) {
+      const percentageShare = BigNumber(rewardUnits.toString()).div(totalEpochRewardUnits.toString())
+      const amount = percentageShare.times(totalDistribution.toString()).toFixed(0)
+
+      distributionsByStakingAddress[address] = {
+        amount,
+        rewardUnits: rewardUnits.toString(),
+        totalRewardUnits: totalRewardUnits.toString(),
+        rewardAddress: runeAddress,
+        txId: '',
+      }
+    }
+
+    return distributionsByStakingAddress
+  }
+
+  async calculateRewards(
+    startBlock: bigint,
+    endBlock: bigint,
+    secondsInEpoch: bigint,
+    totalDistribution: BigNumber,
+  ): Promise<{ totalRewardUnits: string; distributionsByStakingAddress: Record<string, RewardDistribution> }> {
+    const spinner = ora('Calculating reward distribution').start()
+
+    try {
+      const stakeEvents = await this.rpc.getContractEvents({
+        address: ARBITRUM_RFOX_PROXY_CONTRACT_ADDRESS,
+        abi: stakingV1Abi,
+        eventName: 'Stake',
+        fromBlock: 'earliest',
+        toBlock: endBlock,
+      })
+
+      const addresses = [
+        ...new Set(stakeEvents.map(event => event.args.account).filter(address => Boolean(address))),
+      ] as Address[]
+
+      const closingStateByStakingAddress = await this.getClosingStateByStakingAddress(addresses, startBlock, endBlock)
+      const distributionsByStakingAddress = await this.getDistributionsByStakingAddress(
+        closingStateByStakingAddress,
+        totalDistribution,
+      )
+
+      const totalEpochRewardUnits = Object.values(closingStateByStakingAddress).reduce(
+        (prev, { rewardUnits }) => prev + rewardUnits,
+        0n,
+      )
+
+      const totalEpochDistribution = Object.values(distributionsByStakingAddress).reduce(
+        (prev, { amount }) => prev.plus(BigNumber(amount)),
+        BigNumber(0),
+      )
+
+      spinner.succeed()
+
+      info(`Total addresses receiving rewards: ${addresses.length}`)
+
+      const epochRewardUnits = RFOX_REWARD_RATE * secondsInEpoch
+      const epochRewardUnitsMargin = BigNumber(epochRewardUnits.toString()).times(0.01)
```

Comment:
minor: I think we might want to make this smaller.  Its pretty easy to imagine a scenario where a small staker who is less than 1% of everything staked has some sequence of actions that we didn't cover in the previous logic.

If that occurred this check wouldn't catch it.  

REWARD_RATE is 1e27, so any rounding errors due to integer math truncation should be very small.  I think we might choose to be overly cautious and assume rounding errors of < 1 basis point (.01% or .0001) is expected

Reply:
Reduced margin to .01%: fb4227d50539082960b42f2465e5b2c4c2a36fa0